From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sun, 2 Mar 2025 19:58:04 +0100
Subject: [PATCH 085/225] net: wwan: expose Unisoc modem channels
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Expose NV access, timesync, diag and AT command channels provided by the
integrated modem in Unisoc SoCs via the WWAN framework.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/net/wwan/rpmsg_wwan_ctrl.c | 22 ++++++++++++++++++++--
 drivers/net/wwan/wwan_core.c       | 12 ++++++++++++
 include/linux/wwan.h               |  6 ++++++
 3 files changed, 38 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wwan/rpmsg_wwan_ctrl.c b/drivers/net/wwan/rpmsg_wwan_ctrl.c
index 26756ff0e..19e73879f 100644
--- a/drivers/net/wwan/rpmsg_wwan_ctrl.c
+++ b/drivers/net/wwan/rpmsg_wwan_ctrl.c
@@ -3,6 +3,7 @@
 #include <linux/kernel.h>
 #include <linux/mod_devicetable.h>
 #include <linux/module.h>
+#include <linux/of.h>
 #include <linux/platform_device.h>
 #include <linux/rpmsg.h>
 #include <linux/wwan.h>
@@ -113,6 +114,7 @@ static struct device *rpmsg_wwan_find_parent(struct device *dev)
 static int rpmsg_wwan_ctrl_probe(struct rpmsg_device *rpdev)
 {
 	struct rpmsg_wwan_dev *rpwwan;
+	enum wwan_port_type port_type;
 	struct wwan_port *port;
 	struct device *parent;
 
@@ -127,9 +129,14 @@ static int rpmsg_wwan_ctrl_probe(struct rpmsg_device *rpdev)
 	rpwwan->rpdev = rpdev;
 	dev_set_drvdata(&rpdev->dev, rpwwan);
 
+	if (rpdev->dev.of_node)
+		port_type = (uintptr_t)of_device_get_match_data(&rpdev->dev);
+	else
+		port_type = rpdev->id.driver_data;
+
 	/* Register as a wwan port, id.driver_data contains wwan port type */
-	port = wwan_create_port(parent, rpdev->id.driver_data,
-				&rpmsg_wwan_pops, NULL, rpwwan);
+	port = wwan_create_port(parent, port_type, &rpmsg_wwan_pops,
+				NULL, rpwwan);
 	if (IS_ERR(port))
 		return PTR_ERR(port);
 
@@ -145,6 +152,16 @@ static void rpmsg_wwan_ctrl_remove(struct rpmsg_device *rpdev)
 	wwan_remove_port(rpwwan->wwan_port);
 }
 
+static const struct of_device_id rpmsg_wwan_ctrl_of_table[] = {
+	/* RPMSG channels for Unisoc SoCs with integrated LTE modem */
+	{ .compatible = "sprd,spipe-at", .data = (void*)WWAN_PORT_AT },
+	{ .compatible = "sprd,spipe-nv", .data = (void*)WWAN_PORT_SPRDNV },
+	{ .compatible = "sprd,spipe-diag", .data = (void*)WWAN_PORT_SPRDDIAG },
+	{ .compatible = "sprd,spipe-time", .data = (void*)WWAN_PORT_SPRDTIME },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, rpmsg_wwan_ctrl_of_table);
+
 static const struct rpmsg_device_id rpmsg_wwan_ctrl_id_table[] = {
 	/* RPMSG channels for Qualcomm SoCs with integrated modem */
 	{ .name = "DATA5_CNTL", .driver_data = WWAN_PORT_QMI },
@@ -156,6 +173,7 @@ MODULE_DEVICE_TABLE(rpmsg, rpmsg_wwan_ctrl_id_table);
 
 static struct rpmsg_driver rpmsg_wwan_ctrl_driver = {
 	.drv.name = "rpmsg_wwan_ctrl",
+	.drv.of_match_table = rpmsg_wwan_ctrl_of_table,
 	.id_table = rpmsg_wwan_ctrl_id_table,
 	.probe = rpmsg_wwan_ctrl_probe,
 	.remove = rpmsg_wwan_ctrl_remove,
diff --git a/drivers/net/wwan/wwan_core.c b/drivers/net/wwan/wwan_core.c
index ccce2ad74..8abf9f0e5 100644
--- a/drivers/net/wwan/wwan_core.c
+++ b/drivers/net/wwan/wwan_core.c
@@ -350,6 +350,18 @@ static const struct {
 		.devsuf = "mipc",
 	},
 	/* WWAN_PORT_NMEA is exported via the GNSS subsystem */
+	[WWAN_PORT_SPRDNV] = {
+		.name = "SPRDNV",
+		.devsuf = "sprdnv",
+	},
+	[WWAN_PORT_SPRDDIAG] = {
+		.name = "SPRDDIAG",
+		.devsuf = "sprddiag",
+	},
+	[WWAN_PORT_SPRDTIME] = {
+		.name = "SPRDTIME",
+		.devsuf = "sprdtime",
+	},
 };
 
 static ssize_t type_show(struct device *dev, struct device_attribute *attr,
diff --git a/include/linux/wwan.h b/include/linux/wwan.h
index 1e0e2cb53..0db3cb3e5 100644
--- a/include/linux/wwan.h
+++ b/include/linux/wwan.h
@@ -20,6 +20,9 @@
  * @WWAN_PORT_ADB: ADB protocol control
  * @WWAN_PORT_MIPC: MTK MIPC diagnostic interface
  * @WWAN_PORT_NMEA: embedded GNSS receiver with NMEA output
+ * @WWAN_PORT_SPRDNV: Unisoc NV item pipe
+ * @WWAN_PORT_SPRDDIAG: Unisoc diagnostic interface
+ * @WWAN_PORT_SPRDTIME: Unisoc time synchronization pipe
  *
  * @WWAN_PORT_MAX: Highest supported port types
  * @WWAN_PORT_UNKNOWN: Special value to indicate an unknown port type
@@ -36,6 +39,9 @@ enum wwan_port_type {
 	WWAN_PORT_ADB,
 	WWAN_PORT_MIPC,
 	WWAN_PORT_NMEA,
+	WWAN_PORT_SPRDNV,
+	WWAN_PORT_SPRDDIAG,
+	WWAN_PORT_SPRDTIME,
 
 	/* Add new port types above this line */
 
